src/LibraryUtilities.ts

```diff
@@ -566,48 +285,235 @@ export function updateSections(oldLayoutSpecs: any, newLayoutSpecs: any): void {
+                continue;
+            }
+        } else if (node.fullyQualifiedName.startsWith(include.path)) {
+            nodeInculde = true;
+            let prefixIndex = node.fullyQualifiedName.indexOf(prefix);
+            let newName = node.fullyQualifiedName.substring(prefixIndex + prefix.length);
+            buildLibraryItemsFromName(node, parentItem, newName);
         }
-    });
 
-    _.each(unprocessedNodes, function (node) {
-        buildLibraryItemsFromName(node, sectionData)
-    });
+        if (nodeMatch) {
+            let item = new ItemData(fullyQualifiedNameParts[fullyQualifiedNameParts.length - 1]);
+            item.constructFromTypeListNode(node);
+            parentItem.appendChild(item);
+        }
+
+        if (nodeMatch || nodeInculde) {
+            node.processed = true;
+            i--;
+            t++;
+        }
+
+        // Go to next includeInfo and update previousIncludePath, previousStartingIndex and currentStartingIndex
+        if (!(nodeMatch || nodeInculde) || t > typeListNodes.length) {
+            previousIncludePath = include.path;
+            previousStartingIndex = currentStartingIndex;
+            currentStartingIndex = t;
+        }
+    }
+}
+
+export function buildLibrarySectionsFromLayoutSpecs(loadedTypes: any, layoutSpecs: any, defaultSectionStr: string, miscSectionStr: string): ItemData[] {
+
+    let typeListNodes: TypeListNode[] = [];
+    let sectionElements: LayoutElement[] = [];
+
+    // Converting raw data to strongly typed data.
+    for (let loadedType of loadedTypes.loadedTypes) {
+        typeListNodes.push(new TypeListNode(loadedType));
+    }
+
+    for (let section of layoutSpecs.sections) {
+        sectionElements.push(new LayoutElement(section));
+    }
 
-    // Sections other than default section are collapsed by default.
-    sectionData.expanded = false;
+    let includeItemPairs: IncludeItemPair[] = [];
+    let sections = convertLayoutElementToItemData(sectionElements, includeItemPairs);
+    let sortedIncludeItemPairs = includeItemPairs.sort((a, b) => a.include.path.localeCompare(b.include.path));
+    let sortedTypeListNodes = typeListNodes.sort((a, b) => a.fullyQualifiedName.localeCompare(b.fullyQualifiedName));
 
-    return sectionData;
+    if (!isValidIncludeInfo(sortedIncludeItemPairs)) {
+        console.error("Invalid layoutSpecs: include info is not unique");
+    }
+
+    constructFromIncludeInfo(sortedTypeListNodes, sortedIncludeItemPairs);
+
+    // Misc section will take all unprocessed nodes
+    let miscSection = sections.find(section => section.text == miscSectionStr);
+    let unprocessedNodes = sortedTypeListNodes.filter(node => !node.processed);
+    unprocessedNodes.forEach(node => buildLibraryItemsFromName(node, miscSection));
+
+    for (let section of sections) {
+        // Change the itemType of the outermost nodes to category
+        section.childItems.forEach(item => {
+            if (item.itemType == "group") {
```

Comment:
Why do we bother testing the itemType, we can directly assign if the second level items are always of the category type?